Structured kernel definitions

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48116 Structured kernels generate Meta registrations
* #48113 Get TestTorch.test_empty_meta working again
* **#45277 Structured kernel definitions**

Implements structured kernels as per https://github.com/pytorch/rfcs/pull/9 and ports upsample_nearest1d to use the framework.

The general structure of this diff:

- Define a new syntax for specifying structured kernels in `native_functions.yaml`. You put `structured: True` on the `out` function (that's what you implement) and `structured_delegate: foo.out` on the functional/inplace variants to define them in terms of the `out` function. There's a bunch of new consistency checking to see if you've done this right, though the error messages are of varying quality. This is most of what's going on in tools.codegen.model
- NativeFunctionGroup turns into StructuredNativeFunctions. Previously I thought that maybe we would use this grouping mechanism for both structured and unstructured kernels, but it turned out that Jiakai needed to make his own grouping structure. So now I've specialized it for structured kernels, which also means I get to add a bunch of invariants, like requiring structured kernels to have both a functional and an out variant. This is the lower bundle of changes in tools.codegen.model
- When you make an out kernel structured, this induces us to generate a new meta function signature for you to write shape checking and output allocation code. The signatures of these is defined by `tools.codegen.api.meta` and generated into `MetaFunctions.h`. Coverage here is very bare bones and will be driven by actual operators we port as we go.
- The meaty part of code generation is what we do when we have some grouped StructuredNativeFunctions. We continue to generate a wrapper per function type, but they're are a bit different as the call your meta functions, and make reference to the actual implementations in out.
- Then there's a port of `upsample_nearest1d`; easiest to review by just looking at what the final code looks like.

Missing pieces:

- Stride calculation in TensorMeta
- Sufficient sanity checking for inplace/out variants
- Enough rope to make TensorIterator work

This PR improves instruction counts on `upsample_nearest1d` because it eliminates an extra redispatch. Testing `at::upsample_nearest1d(x, {10});`

* Functional: before 1314105, after 1150705
* Out: before 915705, after 838405

These numbers may be jittered up to +-16400 (which is the difference when I tested against an unaffected operator `at::upsample_linear1d`), though that may also because unrelated changes affected all operators globally.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

ezyang committed:
[PROTOTYPE] Structured kernel definitions

See https://fb.quip.com/o4JUAuefYVP3

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ljk53 commented on tools/codegen/model.py:
```diff
@@ -197,8 +204,59 @@ def __post_init__(self) -> None:
 SchemaKind = Enum('SchemaKind', ('functional', 'inplace', 'out'))
 
+@dataclass(frozen=True)
+class NativeFunctionGroup:
```
Thanks for creating this!

ezyang replied:
Yeah, should backport this one to master ASAP

ezyang committed:
Update on "[PROTOTYPE] Structured kernel definitions"


See https://fb.quip.com/o4JUAuefYVP3

This mostly follows the same structure as the proposal, though there have been some short cuts taken. It doesn't currently build because I haven't actually implemented the meta function for the function I marked as structured. Still needs a lot of work, the prototype is here just to show feasibility.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[PROTOTYPE] Structured kernel definitions"


See https://fb.quip.com/o4JUAuefYVP3

This mostly follows the same structure as the proposal, though there have been some short cuts taken. It doesn't currently build because I haven't actually implemented the meta function for the function I marked as structured. Still needs a lot of work, the prototype is here just to show feasibility.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[PROTOTYPE] Structured kernel definitions"


See https://github.com/pytorch/rfcs/pull/9

This mostly follows the same structure as the proposal, though there have been some short cuts taken. It doesn't currently build because I haven't actually implemented the meta function for the function I marked as structured. Still needs a lot of work, the prototype is here just to show feasibility.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

[ghstack-poisoned]

ezyang committed:
Update on "[PROTOTYPE] Structured kernel definitions"


See https://github.com/pytorch/rfcs/pull/9

This mostly follows the same structure as the proposal, though there have been some short cuts taken. It doesn't currently build because I haven't actually implemented the meta function for the function I marked as structured. Still needs a lot of work, the prototype is here just to show feasibility.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

[ghstack-poisoned]

ezyang committed:
Update on "Structured kernel definitions"


Implements structured kernels as per https://github.com/pytorch/rfcs/pull/9
and ports upsample_nearest1d to use the framework.

There is a new meta api which is the calling convention for TensorMeta
calculation functions.  Most of the new codegen lives in
structured_func; check out the RFC for an explanation of what the code
looks like.

Missing pieces:

- Stride calculation in TensorMeta
- Sufficient sanity checking for inplace/out variants
- Enough rope to make TensorIterator work

There's some hacks which I can work harder to unwind:

- I need to get upsample_nearest1d to be registered as abstract: True
  in Declarations.yaml even though it has no dispatch table (as it
  is implicitly filled by upsample_nearest1d.out).  I ended up
  hacking this up by just adding a new field 'abstract: True'
  that lets you manually override the abstractness.  Better would
  be to just teach the codegen to fill this correctly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

[ghstack-poisoned]

ljk53 commented on tools/codegen/gen.py:
```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+    const OptionalDeviceGuard device_guard(device_of({device_of}));
+"""
+
+                return f"""\
```
When I studied the codegen, templates like this helps me learn the connection between internal functions and the final output; but sometimes I felt seeing a little concrete output example could be more intuitive (the downside is that we need remember to update the example). Maybe including a link to the RFC above structured_func()?

ezyang commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -0,0 +1,27 @@
+
+namespace at {
+
+struct TensorMeta {
```
strides will subsume contiguous/channels_last, and will be enough to fully specify the output without getting into non dense tensors. There might be more extensibility here when TensorIterator comes in the mix though.

albanD replied:
Or should we consider this class to replace the existing "collection of fields"  on TensorImpl?

ezyang replied:
I'd prefer to keep them decoupled for now, to make it easier to make representational changes in one or the other.

ailzhang reviewed (commented):
LGTM! The `abstract` hack might be a little confusing to developers, but it's okay if it's temporary :D

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+
+        # Illegal to not have dispatch for structured kernels
+        with native_function_manager(g.out):
+            assert g.out.dispatch is not None, "dispatch on structured kernels currently mandatory"
```
Is there something preventing this currently? Looking at native_functions.yaml, I can see a few ops without dispatch keys that look like they could benefit from structured kernels (e.g. arccosh and arcsinh)

ezyang replied:
This is a good question. I don't know the answer off the top of my head, I'll probably have to figure it out by trying to port arccosh.

bdhirsh replied:
actually... maybe this will get fixed for free from your other PR? If we decide that any ops without dispatch entries will implicitly get a Math key.

ezyang replied:
Well, we certainly don't have to assert dispatch is not None anymore :) But I'm still not sure what will happen with this logic when it is run on a "Math" kernel, since these generate slightly differently than normal kernels.

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
 
-    return func
+    def structured_func(g: NativeFunctionGroup) -> List[str]:
+        assert g.out is not None
```
Would it make sense at some point to impose that all 3 variants of a NativeFunctionGroup exist (or if not, are there explicit reasons to not want an op to have all 3 variants)? Probably not at this point (especially if that's not the case for a lot of ops), but maybe once more/most ops have migrated to structured kernels?

ezyang replied:
I think it makes sense to always require an `_out` variant; we don't require this right now and so there are a bunch of functions that are missing `_out` variants when they should have them.

Inplace variant doesn't make sense to enforce though; many functions can't do inplace at all because their output size doesn't match any of their input sizes.

bhosmer replied:
What about requiring at least `functional`? That seems like the canonical member (at least, I can't think of any cases that lack one). 

If that axiom holds, there's the added benefit of being able to conceptually think of a NFG as a canonical definition + two optional variants... also means no need to do dynamic checking for odd cases like `NativeFunctionGroup(None, None, None)`.

This is kind of monday morning quarterbacking on NFG, sorry about that.

ezyang replied:
In the latest version of the patch these are all required now.

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+            if target is Target.DECLARATION:
+                # Not necessary!  Only TypeDefault is actually used, see
+                # https://github.com/pytorch/pytorch/issues/46319
+                return ""
```
Maybe have subfunc return Optional[str] to fit the pattern of the other compute methods, if there's no semantic difference between empty string and None?

ezyang replied:
Mooted as we no longer gen declarations.

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+                    if local.use_c10_dispatcher() is UseC10Dispatcher.full:
+                        return f'm.impl("{f.func.name}", TORCH_FN({type_name}));'
+                    else:
+                        # SIGHHHHH
```
nit: `# TODO SIGHHHHHHH`

bhosmer reviewed (approved):
A few notes inline but this looks clean and solid to me. I'm glad it stops short of building the `Common` dispatch key setup described in the RFC - definitely have more worries about that than anything here 😁

smessmer commented:
We finally cleaned up some of the old codegen mess and are in the process of cleaning up more, but I don't think that this means codegen is now ok and we can run on and add more and more complexity to it. The new codegen is much easier to understand, and to preserve that we should not add new features to it but we should work on removing more and more features from it until it is even easier to read. Codegen in general is still bad.

The problem with codegen is not only that the old codegen was unreadable, but also that there's an additional system that people need to know about and understand when they want to understand how our system works. And generated code is only present after you built PyTorch, so you can't easily grep for functions, follow definitions in your IDE, or even try to find which file the function you're looking for is in. It's much easier to figure out what your code is doing when there's no codegen involved.

So far, the only software component where we really need codegen, seems to be the C++ frontend, and I thought we had set the vision a while ago of removing all codegen except for the C++ frontend.

@ezyang When you started this work stream, we had talked about templated solutions that did seem to work, why did we switch back to codegen?

ezyang commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -0,0 +1,27 @@
+struct TensorMeta {
+  DimVector sizes;
+  // TODO: DimVector strides;
+  TensorOptions options;
```
Yeah, hoping to flush this and other inefficiencies out when I'm ready for benchmarking.

ezyang replied:
Have a constructor now.

ezyang commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -0,0 +1,27 @@
+namespace at {
+
+struct TensorMeta {
+  DimVector sizes;
```
DimVector isn't limited to rank 5. Rather, it has enough slots to hold five array entries inline without a dynamic allocation; if you spill past that then we'll dynamically allocate a vector and store the sizes in there.

bhosmer replied:
Oh that's right

ezyang commented on aten/src/ATen/native/UpSampleNearest1d.cpp:
```diff
@@ -58,36 +23,50 @@ static void upsample_nearest1d_backward_out_cpu_template(
+  check_dim_size(grad_output, 3, 1, full_output_size[1]);
+  check_dim_size(grad_output, 3, 2, full_output_size[2]);
+
+  return new_meta(grad_output, input_size);
```
I hope there are opportunities for utility functions; need to do more functions to say though.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8323,29 +8321,27 @@
-  dispatch:
-    CPU: upsample_nearest1d_cpu
-    CUDA: upsample_nearest1d_cuda
+  abstract: True  # HACK
```
I don't intend to land the PR with the abstract keyword. Need to do some major pre-surgery to get rid though.

> But I wonder if we won't run into trouble when the naming convention needs to be disrupted for one reason or other.

There's no overload naming convention here; matching of signatures is done purely by looking at the type of the function, and the function name (sans overload name). That being said...

> In that world we might want a field that actually specifies the structured op we want to delegate to - delegate: upsample_nearest_1d.out or something.

I'm OK with doing this. It is technically redundant information but it gives us an opportunity to give better error messages and makes the intent in YAML more clear. That being said...

> opposed to silently generating code that would result in link errors 20 minutes later or whatever :)

I'm pretty sure that in the current formulation if you mismatch the signatures of two functions in the same structured block, you'll get a codegen error (because the dangling signature will fail to find the _out version).

bhosmer replied:
You're absolutely right, the grouping machinery is definitely robust enough to handle all the cases I was thinking about (missing `out` target, multiple clusters in a single op, overload name shenanigans). So I think the only motivation that remains is the thing about improving errors based on their expressed intent.

ezyang replied:
@bhosmer There are also more serious changes we can make to how we setup `native_functions.yaml`; for example, we can force the user to group related signatures. This actually hearkens to my original original proposal (https://fb.quip.com/Z7PYApb0uqwR#VBbACAwaTQ4 FB-only sorry), but people seemed allergic to it the first time around.

bhosmer replied:
> people seemed allergic to it the first time around.

Do you remember what the counterarguments were? I always thought grouping variants together was a good idea - not just for readability in the YAML but also to spare the intake logic from performing exactly that grouping logic (plus sanity checks etc.) behind the scenes. 

Now that structured kernels will be adding an explicit intra-group relationship, having a declaration syntax that puts members together makes even more sense to me. 

I guess one possible issue is if different variants support different dispatch keys. Does that ever happen?

ezyang replied:
For the record, the new version of the patch doesn't group `native_functions.yaml`, but has `structured` and `structured_delegate` to make the structure here more explicit

ezyang commented on tools/codegen/api/meta.py:
```diff
@@ -0,0 +1,59 @@
+#     the tensor in question
+
+def name(f: FunctionSchema) -> str:
+    assert f.name.overload_name == ""
```
In the one example I ported they all have overload names :) This invariant enforces that you "stripped" the overload name using signature() to get at the functional signature.

bhosmer replied:
Ah right - sorry, careless read.

ezyang commented:
@smessmer I think I answered your question in the RFC at https://github.com/pytorch/rfcs/pull/9/files#diff-97e82d5fc6776ce3ff1c347db7b6344e877322602d5a4301a6e6156e489a3190R295

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+
+    def group_func(g: NativeFunctionGroup) -> List[str]:
+        if g.structured():
+            return structured_func(g)
```
IIUC the dispatch key checks [here](https://github.com/pytorch/pytorch/pull/45277/files#diff-6382c847dac2d011f822c492e61f9fa0eceb47c2d6fc5fb4f0c42af726c1a3aaR348-R352) might kick us back out with `[]`, is that ok downstream?

ezyang replied:
Err, yes? The whole point of returning [] sometimes is that sometimes you don't have a registration for a given dispatch key

bhosmer replied:
Ah got it - I was reading "unsupported" in the called function as meaning unsupported by structured kernels specifically - forgot that this gets called for literally all dispatch keys. All caught up now 😁

ezyang commented on tools/codegen/gen.py:
```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+                return ""
+            elif target is Target.DEFINITION:
+                assert dispatch is not None
+                assert g.out is not None  # sigh
```
@bhosmer here's a suggested restructuring: what if we just make classes to represent the closures in question? Then the closed over variables will be explicitly notated as `self.blah` and you can see what all the closed over variables are in one place.

bhosmer replied:
Yeah, definitely seems like that would solve the readability/entropy worry of ambient referents equally well. I'm not sure as an idiom it's any less cumbersome to *write* than non-closure functions with everything explicitly passed in, but I'm definitely interested to see how it turns out, if you're down to give it a spin 😁

ezyang replied:
We fixed it in https://github.com/pytorch/pytorch/pull/47008

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -394,5 +400,21 @@ def from_schema(func: FunctionSchema) -> 'NativeSignature':
+class MetaArgument:
+    type: str
+    name: str
+    # By fiat, meta argument functions must be on full c10 dispach
```
basically, there is always a one-to-one correspondence between JIT argument (Argument) and meta argument. This is true because I am going to explicitly not support structured kernels for function signatures who cannot support `use_c10_dispatcher: full`

bhosmer commented:
Re the [new yaml syntax bakeoff](https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#structured-keyword-proposal) - reiterating a comment above, I'd strongly pitch explicit grouping in the surface syntax, I think it's strictly better. (Which also means I'm completely retracting the 'delegate' variation - it solves a problem that doesn't exist with explicit grouping.)

In the RFC you mention "harder to ingest for people who don't care about grouping" - if you mean humans, I can't imagine any reading mode where the the grouped syntax isn't better - clearer, more informative about the way the system structures things. If you mean software, I'm pretty sure it's easier to ungroup on load than it is to group, like the main codegen script does currently.

As a minor point, I wouldn't introduce a new keyword either. Just `func` with an alternate syntax.

ezyang commented:
For posterity's sake; after @bhosmer's most recent post about the syntax, we discussed various options in our meeting, and concluded that while rounding up all the syntax would make sense, it still should be treated as an orthogonal problem. So I'm going to do some version of https://github.com/pytorch/pytorch/pull/45277#discussion_r512426171 in the short term.

ezyang committed:
Update on "Structured kernel definitions"


Implements structured kernels as per https://github.com/pytorch/rfcs/pull/9
and ports upsample_nearest1d to use the framework.

There is a new meta api which is the calling convention for TensorMeta
calculation functions.  Most of the new codegen lives in
structured_func; check out the RFC for an explanation of what the code
looks like.

Missing pieces:

- Stride calculation in TensorMeta
- Sufficient sanity checking for inplace/out variants
- Enough rope to make TensorIterator work

There's some hacks which I can work harder to unwind:

- I need to get upsample_nearest1d to be registered as abstract: True
  in Declarations.yaml even though it has no dispatch table (as it
  is implicitly filled by upsample_nearest1d.out).  I ended up
  hacking this up by just adding a new field 'abstract: True'
  that lets you manually override the abstractness.  Better would
  be to just teach the codegen to fill this correctly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

[ghstack-poisoned]

ezyang committed:
Update on "Structured kernel definitions"


Implements structured kernels as per https://github.com/pytorch/rfcs/pull/9 and ports upsample_nearest1d to use the framework.

The general structure of this diff:

- Define a new syntax for specifying structured kernels in `native_functions.yaml`. You put `structured: True` on the `out` function (that's what you implement) and `structured_delegate: foo.out` on the functional/inplace variants to define them in terms of the `out` function. There's a bunch of new consistency checking to see if you've done this right, though the error messages are of varying quality. This is most of what's going on in tools.codegen.model
- NativeFunctionGroup turns into StructuredNativeFunctions. Previously I thought that maybe we would use this grouping mechanism for both structured and unstructured kernels, but it turned out that Jiakai needed to make his own grouping structure. So now I've specialized it for structured kernels, which also means I get to add a bunch of invariants, like requiring structured kernels to have both a functional and an out variant. This is the lower bundle of changes in tools.codegen.model
- When you make an out kernel structured, this induces us to generate a new meta function signature for you to write shape checking and output allocation code. The signatures of these is defined by `tools.codegen.api.meta` and generated into `MetaFunctions.h`. Coverage here is very bare bones and will be driven by actual operators we port as we go.
- The meaty part of code generation is what we do when we have some grouped StructuredNativeFunctions. We continue to generate a wrapper per function type, but they're are a bit different as the call your meta functions, and make reference to the actual implementations in out.
- Then there's a port of `upsample_nearest1d`; easiest to review by just looking at what the final code looks like.

Missing pieces:

- Stride calculation in TensorMeta
- Sufficient sanity checking for inplace/out variants
- Enough rope to make TensorIterator work

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

[ghstack-poisoned]

ezyang committed:
Update on "Structured kernel definitions"


Implements structured kernels as per https://github.com/pytorch/rfcs/pull/9 and ports upsample_nearest1d to use the framework.

The general structure of this diff:

- Define a new syntax for specifying structured kernels in `native_functions.yaml`. You put `structured: True` on the `out` function (that's what you implement) and `structured_delegate: foo.out` on the functional/inplace variants to define them in terms of the `out` function. There's a bunch of new consistency checking to see if you've done this right, though the error messages are of varying quality. This is most of what's going on in tools.codegen.model
- NativeFunctionGroup turns into StructuredNativeFunctions. Previously I thought that maybe we would use this grouping mechanism for both structured and unstructured kernels, but it turned out that Jiakai needed to make his own grouping structure. So now I've specialized it for structured kernels, which also means I get to add a bunch of invariants, like requiring structured kernels to have both a functional and an out variant. This is the lower bundle of changes in tools.codegen.model
- When you make an out kernel structured, this induces us to generate a new meta function signature for you to write shape checking and output allocation code. The signatures of these is defined by `tools.codegen.api.meta` and generated into `MetaFunctions.h`. Coverage here is very bare bones and will be driven by actual operators we port as we go.
- The meaty part of code generation is what we do when we have some grouped StructuredNativeFunctions. We continue to generate a wrapper per function type, but they're are a bit different as the call your meta functions, and make reference to the actual implementations in out.
- Then there's a port of `upsample_nearest1d`; easiest to review by just looking at what the final code looks like.

Missing pieces:

- Stride calculation in TensorMeta
- Sufficient sanity checking for inplace/out variants
- Enough rope to make TensorIterator work

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

[ghstack-poisoned]

bhosmer commented on tools/codegen/model.py:
```diff
@@ -210,52 +242,65 @@ def __post_init__(self) -> None:
         out = d.get(SchemaKind.out)
-        return NativeFunctionGroup(
+        if functional is None or out is None or not out.structured:
+            return None
```
Should it be an error to specify out as structured without declaring a functional variant? (Or an assert if it's checked upstream, I guess)

Don't know if there are callers that need this to be permissive, but e.g. the code path that runs through `flatten_pre_group()` looks like it might benefit from error checking.

ezyang replied:
It's checked separately, in `validate_unstructured`. I didn't put error tests here because I don't have error context available.

bhosmer commented on tools/codegen/model.py:
```diff
@@ -112,6 +112,19 @@ class NativeFunction:
     # defined.  This is for conveniently reporting error messages!
     loc: 'Location'
 
+    # Whether or not this out functions is a "structured kernel".  Structured
```
nit: function

bhosmer commented on tools/codegen/model.py:
```diff
@@ -112,6 +112,19 @@ class NativeFunction:
+
+    # Whether or not this non-out function is a structured kernel, defined
+    # in terms of the out kernel referenced by the string here.
+    structured_delegate: Optional['OperatorName']
```
Data modeling OCD obliges me to note the room for nonsense in having both these properties 😁 

In a perfect world we'd have a single property whose type depended on `func.kind`, but in the absence of that, there's no great choice (assuming making `SchemaKind` into an ADT that carries this information directly is either impossible in mypy, or unwanted for some other reason).

Half-hearted pitch for 
```
structured: Optional['OperatorName']
```
and out functions just use `''` or whatever and `is_structured()` just tests `is not None`. `get_structured_delegate()` could do the obvious.

ezyang replied:
Yeah, this choice was driven by what I wanted surface syntax to look like, and then trying to keep NativeFunction class looking similar. What do you think is clearer:

```
- func: foo.out
  structured: None

- func: foo.Tensor
  structured: foo.out
```

or 

```
- func: foo.out
  structured: true

- func: foo.Tensor
  structured_delegate: foo.out
```

bhosmer replied:
#2. I think "`structured: None` means it's structured" would be hella confusing. (More generally I think it's better in the surface syntax to have the attributes be, like, monomorphic :)

Matching representation to surface syntax is a nice through line - might be worth a one-liner in the comment header for `NativeFunction` (apologies if it's there already and I missed it)

bhosmer commented:
> @bhosmer I fixed up all the hacks. 

Looks great! couple drive-by observations but nothing big

ezyang committed:
Update on "Structured kernel definitions"


Implements structured kernels as per https://github.com/pytorch/rfcs/pull/9 and ports upsample_nearest1d to use the framework.

The general structure of this diff:

- Define a new syntax for specifying structured kernels in `native_functions.yaml`. You put `structured: True` on the `out` function (that's what you implement) and `structured_delegate: foo.out` on the functional/inplace variants to define them in terms of the `out` function. There's a bunch of new consistency checking to see if you've done this right, though the error messages are of varying quality. This is most of what's going on in tools.codegen.model
- NativeFunctionGroup turns into StructuredNativeFunctions. Previously I thought that maybe we would use this grouping mechanism for both structured and unstructured kernels, but it turned out that Jiakai needed to make his own grouping structure. So now I've specialized it for structured kernels, which also means I get to add a bunch of invariants, like requiring structured kernels to have both a functional and an out variant. This is the lower bundle of changes in tools.codegen.model
- When you make an out kernel structured, this induces us to generate a new meta function signature for you to write shape checking and output allocation code. The signatures of these is defined by `tools.codegen.api.meta` and generated into `MetaFunctions.h`. Coverage here is very bare bones and will be driven by actual operators we port as we go.
- The meaty part of code generation is what we do when we have some grouped StructuredNativeFunctions. We continue to generate a wrapper per function type, but they're are a bit different as the call your meta functions, and make reference to the actual implementations in out.
- Then there's a port of `upsample_nearest1d`; easiest to review by just looking at what the final code looks like.

Missing pieces:

- Stride calculation in TensorMeta
- Sufficient sanity checking for inplace/out variants
- Enough rope to make TensorIterator work

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24253555](https://our.internmc.facebook.com/intern/diff/D24253555)

[ghstack-poisoned]
